### ISSUE TITLE
aasm 4.11.1 & RAILS 4.2.0: NoMethodError (undefined method `aasm_state' ....

### ISSUE THREAD

**Issue by emclab** [aasm/aasm#443]
itemlized_spendingx is an engine and spending is the table name. Instead of using default 'aasm_state', column 'wf_state'  is used in table spending. When saving a new spending, there is an error of above. The controller code which cause the error is:

 def create
      @spending = ItemlizedSpendingx::Spending.new(new_params)
      if @spending.save   #<<<<<<<=======causing error
          redirect_to URI.escape(SUBURI + "/view_handler?index=0&msg=Successfully Saved!")
       end
    end

Here is the full error trace:

NoMethodError (undefined method `aasm_state' for #<ItemlizedSpendingx::Spending:0x007f3e359bd988>):
  activemodel (4.2.0) lib/active_model/attribute_methods.rb:433:in `method_missing'
  aasm (4.11.1) lib/aasm/persistence/base.rb:36:in `aasm_read_state'
  aasm (4.11.1) lib/aasm/instance_base.rb:12:in `current_state'
  aasm (4.11.1) lib/aasm/persistence/active_record_persistence.rb:233:in `aasm_invalid_state?'
  aasm (4.11.1) lib/aasm/persistence/active_record_persistence.rb:225:in `block in aasm_validate_states'
  aasm (4.11.1) lib/aasm/persistence/active_record_persistence.rb:223:in `each'
  aasm (4.11.1) lib/aasm/persistence/active_record_persistence.rb:223:in `aasm_validate_states'
  activesupport (4.2.0) lib/active_support/callbacks.rb:427:in `block in make_lambda'
  activesupport (4.2.0) lib/active_support/callbacks.rb:189:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:189:in `block in simple'
  activesupport (4.2.0) lib/active_support/callbacks.rb:92:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:92:in `_run_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:734:in `_run_validate_callbacks'
  activemodel (4.2.0) lib/active_model/validations.rb:395:in `run_validations!'
  activemodel (4.2.0) lib/active_model/validations/callbacks.rb:113:in `block in run_validations!'
  activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `_run_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:734:in `_run_validation_callbacks'
  activemodel (4.2.0) lib/active_model/validations/callbacks.rb:113:in `run_validations!'
  activemodel (4.2.0) lib/active_model/validations.rb:334:in `valid?'
  activerecord (4.2.0) lib/active_record/validations.rb:58:in `valid?'
  activerecord (4.2.0) lib/active_record/validations.rb:83:in `perform_validations'
  activerecord (4.2.0) lib/active_record/validations.rb:37:in `save'
  activerecord (4.2.0) lib/active_record/attribute_methods/dirty.rb:21:in `save'
  activerecord (4.2.0) lib/active_record/transactions.rb:286:in `block (2 levels) in save'
  activerecord (4.2.0) lib/active_record/transactions.rb:347:in `block in with_transaction_returning_status'
  activerecord (4.2.0) lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `block in transaction'
  activerecord (4.2.0) lib/active_record/connection_adapters/abstract/transaction.rb:188:in `within_new_transaction'
  activerecord (4.2.0) lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `transaction'
  activerecord (4.2.0) lib/active_record/transactions.rb:220:in `transaction'
  activerecord (4.2.0) lib/active_record/transactions.rb:344:in `with_transaction_returning_status'
  activerecord (4.2.0) lib/active_record/transactions.rb:286:in `block in save'
  activerecord (4.2.0) lib/active_record/transactions.rb:301:in `rollback_active_record_state!'
  activerecord (4.2.0) lib/active_record/transactions.rb:285:in `save'
  /home/ubuntu/.rvm/gems/ruby-2.2.6/bundler/gems/itemlized_spendingx-c05187165a46/app/controllers/itemlized_spendingx/spendings_controller.rb:36:in `create'
  actionpack (4.2.0) lib/action_controller/metal/implicit_render.rb:4:in `send_action'
  actionpack (4.2.0) lib/abstract_controller/base.rb:198:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/rendering.rb:10:in `process_action'
  actionpack (4.2.0) lib/abstract_controller/callbacks.rb:20:in `block in process_action'
  activesupport (4.2.0) lib/active_support/callbacks.rb:117:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:117:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:219:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:219:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:219:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:92:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:92:in `_run_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:734:in `_run_process_action_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:81:in `run_callbacks'
  actionpack (4.2.0) lib/abstract_controller/callbacks.rb:19:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/rescue.rb:29:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/instrumentation.rb:31:in `block in process_action'
  activesupport (4.2.0) lib/active_support/notifications.rb:164:in `block in instrument'
  activesupport (4.2.0) lib/active_support/notifications/instrumenter.rb:20:in `instrument'
  activesupport (4.2.0) lib/active_support/notifications.rb:164:in `instrument'
  actionpack (4.2.0) lib/action_controller/metal/instrumentation.rb:30:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/params_wrapper.rb:250:in `process_action'
  activerecord (4.2.0) lib/active_record/railties/controller_runtime.rb:18:in `process_action'
  actionpack (4.2.0) lib/abstract_controller/base.rb:137:in `process'
  actionview (4.2.0) lib/action_view/rendering.rb:30:in `process'
  actionpack (4.2.0) lib/action_controller/metal.rb:195:in `dispatch'
  actionpack (4.2.0) lib/action_controller/metal/rack_delegation.rb:13:in `dispatch'
  actionpack (4.2.0) lib/action_controller/metal.rb:236:in `block in action'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:73:in `call'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:73:in `dispatch'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:42:in `serve'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:43:in `block in serve'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:30:in `each'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:30:in `serve'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:802:in `call'
  railties (4.2.0) lib/rails/engine.rb:518:in `call'
  railties (4.2.0) lib/rails/railtie.rb:194:in `public_send'
  railties (4.2.0) lib/rails/railtie.rb:194:in `method_missing'
  actionpack (4.2.0) lib/action_dispatch/routing/mapper.rb:51:in `serve'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:43:in `block in serve'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:30:in `each'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:30:in `serve'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:802:in `call'
  rack (1.6.5) lib/rack/etag.rb:24:in `call'
  rack (1.6.5) lib/rack/conditionalget.rb:38:in `call'
  rack (1.6.5) lib/rack/head.rb:13:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/params_parser.rb:27:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/flash.rb:260:in `call'
  rack (1.6.5) lib/rack/session/abstract/id.rb:225:in `context'
  rack (1.6.5) lib/rack/session/abstract/id.rb:220:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/cookies.rb:560:in `call'
  activerecord (4.2.0) lib/active_record/query_cache.rb:36:in `call'
  activerecord (4.2.0) lib/active_record/connection_adapters/abstract/connection_pool.rb:647:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `_run_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:734:in `_run_call_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:81:in `run_callbacks'
  actionpack (4.2.0) lib/action_dispatch/middleware/callbacks.rb:27:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/remote_ip.rb:78:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
  railties (4.2.0) lib/rails/rack/logger.rb:38:in `call_app'
  railties (4.2.0) lib/rails/rack/logger.rb:20:in `block in call'
  activesupport (4.2.0) lib/active_support/tagged_logging.rb:68:in `block in tagged'
  activesupport (4.2.0) lib/active_support/tagged_logging.rb:26:in `tagged'
  activesupport (4.2.0) lib/active_support/tagged_logging.rb:68:in `tagged'
  railties (4.2.0) lib/rails/rack/logger.rb:20:in `call'
  request_store (1.3.2) lib/request_store/middleware.rb:9:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/request_id.rb:21:in `call'
  rack (1.6.5) lib/rack/methodoverride.rb:22:in `call'
  rack (1.6.5) lib/rack/runtime.rb:18:in `call'
  activesupport (4.2.0) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
  rack (1.6.5) lib/rack/sendfile.rb:113:in `call'
  railties (4.2.0) lib/rails/engine.rb:518:in `call'
  railties (4.2.0) lib/rails/application.rb:164:in `call'
  railties (4.2.0) lib/rails/railtie.rb:194:in `public_send'
  railties (4.2.0) lib/rails/railtie.rb:194:in `method_missing'
  passenger (5.0.30) src/ruby_supportlib/phusion_passenger/rack/thread_handler_extension.rb:97:in `process_request'
  passenger (5.0.30) src/ruby_supportlib/phusion_passenger/request_handler/thread_handler.rb:160:in `accept_and_process_next_request'
  passenger (5.0.30) src/ruby_supportlib/phusion_passenger/request_handler/thread_handler.rb:113:in `main_loop'
  passenger (5.0.30) src/ruby_supportlib/phusion_passenger/request_handler.rb:416:in `block (3 levels) in start_threads'
  passenger (5.0.30) src/ruby_supportlib/phusion_passenger/utils.rb:113:in `block in create_thread_and_abort_on_exception'

Something puzzles me is that the error disappears after restarting the web server. Also when I just login to the system, there is no error. However if I leave the system idle for 20-30 min and come back, then the error shows up. My understanding of the error is that the aasm gem is still looking for default column name 'aasm_state'. I can not recall the similar error for previous version (before 4.11). Is it a bug in aasm?

**Comment by emclab** [comment]
Also our aasm do loop is a little complex. Workflow name is used. Here is how the wf def is loaded in "spending" model:

class Spending < ActiveRecord::Base
    include AASM

    def self.load_wf_spec(wf_spec, wf_def_name)
        eval("aasm(:#{wf_def_name}, :column => 'wf_state')  {#{wf_spec}}")
    end
end

method "load_wf_spec" is called in before_action in spendings controller. 

Here  wf_def_name is a string 'basic', wf_spec is a string of full workflow definition as follows:

        " state :initial_state, :initial => true
           state :ceo_reviewing 
           state :approved
           state :rejected
        event :submit_test do
          transitions :from => :initial_state, :to => :ceo_reviewing
        end
        event :ceo_approve_test do
          transitions :from => :ceo_reviewing, :to => :approved
        end 
        event :ceo_reject_test do
          transitions :from => :ceo_reviewing, :to => :rejected
        end "

**Comment by emclab** [comment]
Here is the code in base.rb in aasm:
     def aasm_read_state(name=:default)
        state = send(self.class.aasm(name).attribute_name)  #<<<<< name is 'test'. if self.class.aasm(name).attribute_name returns 'aasm_state', then raise the error. It should return 'wf_state'
        if new_record?
          state.blank? ? aasm(name).determine_state_name(self.class.aasm(name).initial_state) : state.to_sym
        else
          state.blank? ? nil : state.to_sym
        end
      end

**Comment by reidmorrison** [comment]
Does this issue occur in your tests?
Is this a failure that happens every time, or just under load?

Using Markdown on the comments above would make them easier to read: https://guides.github.com/features/mastering-markdown/

**Comment by emclab** [comment]
@reidmorrison The problem happens in production and happens only a few times. The problem is that I don't  know when it will happen.

**Comment by reidmorrison** [comment]
Sounds like it could be a concurrency issue, can you try running against master and see if that helps. I do have one more concurrency fix that I forgot about, and will get that into a PR soon.

**Comment by emclab** [comment]
Just changed the gemfile and pointing aasm to master branch. Let's cross finger.

**Comment by anilmaurya** [comment]
@emclab I just merged another pull request solving concurrent issue, Please bundle update aasm .

I will release new AASM version sooner.

**Comment by emclab** [comment]
@anilmaurya , just updated aasm (gem 'aasm', :git => 'git@github.com:aasm/aasm.git', :branch => 'master'). Please let me know when the PR is available. Thanks.

**Comment by emclab** [comment]
@anilmaurya @reidmorrison what's the issue fixed? Please drop a few lines about the issue. Thanks.

**Comment by emclab** [comment]
So far so good for last few days.

**Comment by anilmaurya** [comment]
@emclab  concurrency issue is resolved in master branch, Looks like your issue is related to concurrency .

**Comment by anilmaurya** [comment]
Just released `AASM-4.12.0` .

**Comment by reidmorrison** [comment]
Thank you @anilmaurya :tada:

**Comment by emclab** [comment]
oops, it strikes again this morning. My version is still at 4.11.1.  Do I need upgrade to 4.12.0 to solve the issue?

**Comment by emclab** [comment]
4.12.0 has the same problem. Here is the error:

F, [2017-03-28T11:44:18.714733 #9504] FATAL -- :
NoMethodError (undefined method `aasm_state' for #<ItemlizedSpendingx::Spending:0x007f207fdbf9b0>):
  activemodel (4.2.0) lib/active_model/attribute_methods.rb:433:in `method_missing'
  aasm (4.12.0) lib/aasm/persistence/base.rb:36:in `aasm_read_state'
  aasm (4.12.0) lib/aasm/instance_base.rb:12:in `current_state'
  aasm (4.12.0) lib/aasm/persistence/active_record_persistence.rb:233:in `aasm_invalid_state?'
  aasm (4.12.0) lib/aasm/persistence/active_record_persistence.rb:225:in `block in aasm_validate_states'
  aasm (4.12.0) lib/aasm/persistence/active_record_persistence.rb:223:in `each'
  aasm (4.12.0) lib/aasm/persistence/active_record_persistence.rb:223:in `aasm_validate_states'
  activesupport (4.2.0) lib/active_support/callbacks.rb:427:in `block in make_lambda'
  activesupport (4.2.0) lib/active_support/callbacks.rb:189:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:189:in `block in simple'
  activesupport (4.2.0) lib/active_support/callbacks.rb:92:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:92:in `_run_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:734:in `_run_validate_callbacks'
  activemodel (4.2.0) lib/active_model/validations.rb:395:in `run_validations!'
  activemodel (4.2.0) lib/active_model/validations/callbacks.rb:113:in `block in run_validations!'
  activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `_run_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:734:in `_run_validation_callbacks'
  activemodel (4.2.0) lib/active_model/validations/callbacks.rb:113:in `run_validations!'
  activemodel (4.2.0) lib/active_model/validations.rb:334:in `valid?'
  activerecord (4.2.0) lib/active_record/validations.rb:58:in `valid?'
  activerecord (4.2.0) lib/active_record/validations.rb:83:in `perform_validations'
  activerecord (4.2.0) lib/active_record/validations.rb:37:in `save'
  activerecord (4.2.0) lib/active_record/attribute_methods/dirty.rb:21:in `save'
  activerecord (4.2.0) lib/active_record/transactions.rb:286:in `block (2 levels) in save'
  activerecord (4.2.0) lib/active_record/transactions.rb:347:in `block in with_transaction_returning_status'
  activerecord (4.2.0) lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `block in transaction'
  activerecord (4.2.0) lib/active_record/connection_adapters/abstract/transaction.rb:188:in `within_new_transaction'
  activerecord (4.2.0) lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `transaction'
  activerecord (4.2.0) lib/active_record/transactions.rb:220:in `transaction'
  activerecord (4.2.0) lib/active_record/transactions.rb:344:in `with_transaction_returning_status'
  activerecord (4.2.0) lib/active_record/transactions.rb:286:in `block in save'
  activerecord (4.2.0) lib/active_record/transactions.rb:301:in `rollback_active_record_state!'
  activerecord (4.2.0) lib/active_record/transactions.rb:285:in `save'
  /home/ubuntu/.rvm/gems/ruby-2.2.6/bundler/gems/itemlized_spendingx-ed89f131fa6c/app/controllers/itemlized_spendingx/spendings_controller.rb:38:in `create'
  actionpack (4.2.0) lib/action_controller/metal/implicit_render.rb:4:in `send_action'
  actionpack (4.2.0) lib/abstract_controller/base.rb:198:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/rendering.rb:10:in `process_action'
  actionpack (4.2.0) lib/abstract_controller/callbacks.rb:20:in `block in process_action'
  activesupport (4.2.0) lib/active_support/callbacks.rb:117:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:117:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:219:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:219:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:219:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:151:in `block in halting_and_conditional'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
  activesupport (4.2.0) lib/active_support/callbacks.rb:92:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:92:in `_run_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:734:in `_run_process_action_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:81:in `run_callbacks'
  actionpack (4.2.0) lib/abstract_controller/callbacks.rb:19:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/rescue.rb:29:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/instrumentation.rb:31:in `block in process_action'
  activesupport (4.2.0) lib/active_support/notifications.rb:164:in `block in instrument'
  activesupport (4.2.0) lib/active_support/notifications/instrumenter.rb:20:in `instrument'
  activesupport (4.2.0) lib/active_support/notifications.rb:164:in `instrument'
  actionpack (4.2.0) lib/action_controller/metal/instrumentation.rb:30:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/params_wrapper.rb:250:in `process_action'
  activerecord (4.2.0) lib/active_record/railties/controller_runtime.rb:18:in `process_action'
  actionpack (4.2.0) lib/abstract_controller/base.rb:137:in `process'
  actionview (4.2.0) lib/action_view/rendering.rb:30:in `process'
  actionpack (4.2.0) lib/action_controller/metal.rb:195:in `dispatch'
  actionpack (4.2.0) lib/action_controller/metal/rack_delegation.rb:13:in `dispatch'
  actionpack (4.2.0) lib/action_controller/metal.rb:236:in `block in action'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:73:in `call'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:73:in `dispatch'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:42:in `serve'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:43:in `block in serve'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:30:in `each'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:30:in `serve'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:802:in `call'
  railties (4.2.0) lib/rails/engine.rb:518:in `call'
  railties (4.2.0) lib/rails/railtie.rb:194:in `public_send'
  railties (4.2.0) lib/rails/railtie.rb:194:in `method_missing'
  actionpack (4.2.0) lib/action_dispatch/routing/mapper.rb:51:in `serve'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:43:in `block in serve'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:30:in `each'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:30:in `serve'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:802:in `call'
  rack (1.6.4) lib/rack/etag.rb:24:in `call'
  rack (1.6.4) lib/rack/conditionalget.rb:38:in `call'
  rack (1.6.4) lib/rack/head.rb:13:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/params_parser.rb:27:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/flash.rb:260:in `call'
  rack (1.6.4) lib/rack/session/abstract/id.rb:225:in `context'
  rack (1.6.4) lib/rack/session/abstract/id.rb:220:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/cookies.rb:560:in `call'
  activerecord (4.2.0) lib/active_record/query_cache.rb:36:in `call'
  activerecord (4.2.0) lib/active_record/connection_adapters/abstract/connection_pool.rb:647:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `_run_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:734:in `_run_call_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:81:in `run_callbacks'
  actionpack (4.2.0) lib/action_dispatch/middleware/callbacks.rb:27:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/remote_ip.rb:78:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
  railties (4.2.0) lib/rails/rack/logger.rb:38:in `call_app'
  railties (4.2.0) lib/rails/rack/logger.rb:20:in `block in call'
  activesupport (4.2.0) lib/active_support/tagged_logging.rb:68:in `block in tagged'
  activesupport (4.2.0) lib/active_support/tagged_logging.rb:26:in `tagged'
  activesupport (4.2.0) lib/active_support/tagged_logging.rb:68:in `tagged'
  railties (4.2.0) lib/rails/rack/logger.rb:20:in `call'
  request_store (1.3.1) lib/request_store/middleware.rb:9:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/request_id.rb:21:in `call'
  rack (1.6.4) lib/rack/methodoverride.rb:22:in `call'
  rack (1.6.4) lib/rack/runtime.rb:18:in `call'
  activesupport (4.2.0) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
  rack (1.6.4) lib/rack/sendfile.rb:113:in `call'
  railties (4.2.0) lib/rails/engine.rb:518:in `call'
  railties (4.2.0) lib/rails/application.rb:164:in `call'
  railties (4.2.0) lib/rails/railtie.rb:194:in `public_send'
  railties (4.2.0) lib/rails/railtie.rb:194:in `method_missing'
  passenger (5.0.30) src/ruby_supportlib/phusion_passenger/rack/thread_handler_extension.rb:97:in `process_request'
  passenger (5.0.30) src/ruby_supportlib/phusion_passenger/request_handler/thread_handler.rb:160:in `accept_and_process_next_request'
  passenger (5.0.30) src/ruby_supportlib/phusion_passenger/request_handler/thread_handler.rb:113:in `main_loop'
  passenger (5.0.30) src/ruby_supportlib/phusion_passenger/request_handler.rb:416:in `block (3 levels) in start_threads'
  passenger (5.0.30) src/ruby_supportlib/phusion_passenger/utils.rb:113:in `block in create_thread_and_abort_on_exception'

**Comment by rpearce** [comment]
I'm seeing this, too, when running `db:create db:migrate db:seed` – where my seeds now have a `User` using AASM being created

**Comment by reidmorrison** [comment]
We are running Rails 4.2 with AASM 4.12 in a high volume, high concurrency environment and have not seen the above issue yet.

To confirm, the following Rails production settings have been set?
~~~ruby
    config.cache_classes = true
    config.eager_load = true
~~~

If you are still seeing the issue then I would suggest adding more relevant data to the issue so that we can see what is different that could be causing the issue. For example, ruby version, formatted stack traces, etc.

Only thing I can think of right now is that the ActiveRecord model has not been loaded yet while another thread is trying to access the `aasm_state` attribute.

We saw issues with other missing AR attributes, so added the following Rails initializer to pre-load all the AR models before starting threads.

~~~ruby
# Define all the dynamic methods/attributes for every ActiveRecord Model
Rails.application.config.after_initialize do |app|
  if app.config.eager_load
    # Create the attribute methods for every AR model before starting any threads
    ActiveRecord::Base.descendants.each do |klass|
      klass.define_attribute_methods rescue nil
    end
  end
end
~~~

**Comment by teyo-soler** [comment]
I have the same issue. Tried to downgrade to 4.10 but getting the same error.

`undefined` method `aasm_state' for #<Contract:0x0055aa9c631950>

`activemodel (5.0.2) lib/active_model/attribute_methods.rb:433:in `method_missing'
aasm (4.10.1) lib/aasm/persistence/base.rb:36:in `aasm_read_state'
aasm (4.10.1) lib/aasm/instance_base.rb:12:in `current_state'
aasm (4.10.1) lib/aasm/persistence/active_record_persistence.rb:224:in `aasm_invalid_state?'
aasm (4.10.1) lib/aasm/persistence/active_record_persistence.rb:216:in `block in aasm_validate_states'
aasm (4.10.1) lib/aasm/persistence/active_record_persistence.rb:214:in `each'
aasm (4.10.1) lib/aasm/persistence/active_record_persistence.rb:214:in `aasm_validate_states'
activesupport (5.0.2) lib/active_support/callbacks.rb:382:in `block in make_lambda'
activesupport (5.0.2) lib/active_support/callbacks.rb:169:in `block (2 levels) in halting'
activesupport (5.0.2) lib/active_support/callbacks.rb:547:in `block (2 levels) in default_terminator'
activesupport (5.0.2) lib/active_support/callbacks.rb:546:in `catch'
activesupport (5.0.2) lib/active_support/callbacks.rb:546:in `block in default_terminator'
activesupport (5.0.2) lib/active_support/callbacks.rb:170:in `block in halting'
activesupport (5.0.2) lib/active_support/callbacks.rb:454:in `block in call'
activesupport (5.0.2) lib/active_support/callbacks.rb:454:in `each'
activesupport (5.0.2) lib/active_support/callbacks.rb:454:in `call'
activesupport (5.0.2) lib/active_support/callbacks.rb:101:in `__run_callbacks__'
activesupport (5.0.2) lib/active_support/callbacks.rb:750:in `_run_validate_callbacks'
activemodel (5.0.2) lib/active_model/validations.rb:408:in `run_validations!'
activemodel (5.0.2) lib/active_model/validations/callbacks.rb:113:in `block in run_validations!'
activesupport (5.0.2) lib/active_support/callbacks.rb:126:in `call'
activesupport (5.0.2) lib/active_support/callbacks.rb:506:in `block (2 levels) in compile'
activesupport (5.0.2) lib/active_support/callbacks.rb:455:in `call'
activesupport (5.0.2) lib/active_support/callbacks.rb:101:in `__run_callbacks__'
activesupport (5.0.2) lib/active_support/callbacks.rb:750:in `_run_validation_callbacks'
activemodel (5.0.2) lib/active_model/validations/callbacks.rb:113:in `run_validations!'
activemodel (5.0.2) lib/active_model/validations.rb:338:in `valid?'
activerecord (5.0.2) lib/active_record/validations.rb:65:in `valid?'
activerecord (5.0.2) lib/active_record/validations.rb:82:in `perform_validations'
activerecord (5.0.2) lib/active_record/validations.rb:44:in `save'
activerecord (5.0.2) lib/active_record/attribute_methods/dirty.rb:22:in `save'
activerecord (5.0.2) lib/active_record/transactions.rb:319:in `block (2 levels) in save'
activerecord (5.0.2) lib/active_record/transactions.rb:395:in `block in with_transaction_returning_status'
activerecord (5.0.2) lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `block in transaction'
activerecord (5.0.2) lib/active_record/connection_adapters/abstract/transaction.rb:189:in `within_new_transaction'
activerecord (5.0.2) lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `transaction'
activerecord (5.0.2) lib/active_record/transactions.rb:211:in `transaction'
activerecord (5.0.2) lib/active_record/transactions.rb:392:in `with_transaction_returning_status'
activerecord (5.0.2) lib/active_record/transactions.rb:319:in `block in save'
activerecord (5.0.2) lib/active_record/transactions.rb:334:in `rollback_active_record_state!'
activerecord (5.0.2) lib/active_record/transactions.rb:318:in `save'
activerecord (5.0.2) lib/active_record/suppressor.rb:41:in `save'
actionpack (5.0.2) lib/action_controller/metal/mime_responds.rb:195:in `respond_to'
actionpack (5.0.2) lib/action_controller/metal/basic_implicit_render.rb:4:in `send_action'
actionpack (5.0.2) lib/abstract_controller/base.rb:188:in `process_action'
actionpack (5.0.2) lib/action_controller/metal/rendering.rb:30:in `process_action'
actionpack (5.0.2) lib/abstract_controller/callbacks.rb:20:in `block in process_action'
activesupport (5.0.2) lib/active_support/callbacks.rb:126:in `call'
activesupport (5.0.2) lib/active_support/callbacks.rb:506:in `block (2 levels) in compile'
activesupport (5.0.2) lib/active_support/callbacks.rb:455:in `call'
activesupport (5.0.2) lib/active_support/callbacks.rb:448:in `block (2 levels) in around'
activesupport (5.0.2) lib/active_support/callbacks.rb:286:in `block (2 levels) in halting'
activesupport (5.0.2) lib/active_support/callbacks.rb:382:in `block in make_lambda'
activesupport (5.0.2) lib/active_support/callbacks.rb:285:in `block in halting'
activesupport (5.0.2) lib/active_support/callbacks.rb:447:in `block in around'
activesupport (5.0.2) lib/active_support/callbacks.rb:455:in `call'
activesupport (5.0.2) lib/active_support/callbacks.rb:448:in `block (2 levels) in around'
activesupport (5.0.2) lib/active_support/callbacks.rb:286:in `block (2 levels) in halting'
/home/teyo/.rvm/gems/ruby-2.4.0/bundler/gems/audited-615627f08c65/lib/audited/sweeper.rb:14:in `around'
activesupport (5.0.2) lib/active_support/callbacks.rb:405:in `public_send'
activesupport (5.0.2) lib/active_support/callbacks.rb:405:in `block in make_lambda'
activesupport (5.0.2) lib/active_support/callbacks.rb:285:in `block in halting'
activesupport (5.0.2) lib/active_support/callbacks.rb:447:in `block in around'
activesupport (5.0.2) lib/active_support/callbacks.rb:455:in `call'
activesupport (5.0.2) lib/active_support/callbacks.rb:448:in `block (2 levels) in around'
activesupport (5.0.2) lib/active_support/callbacks.rb:286:in `block (2 levels) in halting'
/home/teyo/.rvm/gems/ruby-2.4.0/bundler/gems/audited-615627f08c65/lib/audited/sweeper.rb:14:in `around'
activesupport (5.0.2) lib/active_support/callbacks.rb:405:in `public_send'
activesupport (5.0.2) lib/active_support/callbacks.rb:405:in `block in make_lambda'
activesupport (5.0.2) lib/active_support/callbacks.rb:285:in `block in halting'
activesupport (5.0.2) lib/active_support/callbacks.rb:447:in `block in around'
activesupport (5.0.2) lib/active_support/callbacks.rb:455:in `call'
activesupport (5.0.2) lib/active_support/callbacks.rb:101:in `__run_callbacks__'
activesupport (5.0.2) lib/active_support/callbacks.rb:750:in `_run_process_action_callbacks'
activesupport (5.0.2) lib/active_support/callbacks.rb:90:in `run_callbacks'
actionpack (5.0.2) lib/abstract_controller/callbacks.rb:19:in `process_action'
actionpack (5.0.2) lib/action_controller/metal/rescue.rb:20:in `process_action'
actionpack (5.0.2) lib/action_controller/metal/instrumentation.rb:32:in `block in process_action'
activesupport (5.0.2) lib/active_support/notifications.rb:164:in `block in instrument'
activesupport (5.0.2) lib/active_support/notifications/instrumenter.rb:21:in `instrument'
activesupport (5.0.2) lib/active_support/notifications.rb:164:in `instrument'
actionpack (5.0.2) lib/action_controller/metal/instrumentation.rb:30:in `process_action'
actionpack (5.0.2) lib/action_controller/metal/params_wrapper.rb:248:in `process_action'
activerecord (5.0.2) lib/active_record/railties/controller_runtime.rb:18:in `process_action'
actionpack (5.0.2) lib/abstract_controller/base.rb:126:in `process'
actionview (5.0.2) lib/action_view/rendering.rb:30:in `process'
actionpack (5.0.2) lib/action_controller/metal.rb:190:in `dispatch'
actionpack (5.0.2) lib/action_controller/metal.rb:262:in `dispatch'
actionpack (5.0.2) lib/action_dispatch/routing/route_set.rb:50:in `dispatch'
actionpack (5.0.2) lib/action_dispatch/routing/route_set.rb:32:in `serve'
actionpack (5.0.2) lib/action_dispatch/journey/router.rb:39:in `block in serve'
actionpack (5.0.2) lib/action_dispatch/journey/router.rb:26:in `each'
actionpack (5.0.2) lib/action_dispatch/journey/router.rb:26:in `serve'
actionpack (5.0.2) lib/action_dispatch/routing/route_set.rb:725:in `call'
rack (2.0.1) lib/rack/etag.rb:25:in `call'
rack (2.0.1) lib/rack/conditional_get.rb:38:in `call'
rack (2.0.1) lib/rack/head.rb:12:in `call'
rack (2.0.1) lib/rack/session/abstract/id.rb:222:in `context'
rack (2.0.1) lib/rack/session/abstract/id.rb:216:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/cookies.rb:613:in `call'
activerecord (5.0.2) lib/active_record/migration.rb:553:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/callbacks.rb:38:in `block in call'
activesupport (5.0.2) lib/active_support/callbacks.rb:97:in `__run_callbacks__'
activesupport (5.0.2) lib/active_support/callbacks.rb:750:in `_run_call_callbacks'
activesupport (5.0.2) lib/active_support/callbacks.rb:90:in `run_callbacks'
actionpack (5.0.2) lib/action_dispatch/middleware/callbacks.rb:36:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/executor.rb:12:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/remote_ip.rb:79:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/debug_exceptions.rb:49:in `call'
web-console (3.4.0) lib/web_console/middleware.rb:135:in `call_app'
web-console (3.4.0) lib/web_console/middleware.rb:28:in `block in call'
web-console (3.4.0) lib/web_console/middleware.rb:18:in `catch'
web-console (3.4.0) lib/web_console/middleware.rb:18:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
railties (5.0.2) lib/rails/rack/logger.rb:36:in `call_app'
railties (5.0.2) lib/rails/rack/logger.rb:24:in `block in call'
activesupport (5.0.2) lib/active_support/tagged_logging.rb:69:in `block in tagged'
activesupport (5.0.2) lib/active_support/tagged_logging.rb:26:in `tagged'
activesupport (5.0.2) lib/active_support/tagged_logging.rb:69:in `tagged'
railties (5.0.2) lib/rails/rack/logger.rb:24:in `call'
sprockets-rails (3.2.0) lib/sprockets/rails/quiet_assets.rb:13:in `call'
request_store (1.3.2) lib/request_store/middleware.rb:9:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/request_id.rb:24:in `call'
rack (2.0.1) lib/rack/method_override.rb:22:in `call'
rack (2.0.1) lib/rack/runtime.rb:22:in `call'
activesupport (5.0.2) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/executor.rb:12:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/static.rb:136:in `call'
rack (2.0.1) lib/rack/sendfile.rb:111:in `call'
railties (5.0.2) lib/rails/engine.rb:522:in `call'
puma (3.8.2) lib/puma/configuration.rb:224:in `call'
puma (3.8.2) lib/puma/server.rb:600:in `handle_request'
puma (3.8.2) lib/puma/server.rb:435:in `process_client'
puma (3.8.2) lib/puma/server.rb:299:in `block in run'
puma (3.8.2) lib/puma/thread_pool.rb:120:in `block in spawn_thread'
`

**Comment by evserykh** [comment]
Same problem with Rails 5.1.1

**Comment by emclab** [comment]
The problem shows up again after new deployment. The error msg is exactly as what described above.  I have ruby 2.2.6 on production.

**Comment by skorth** [comment]
Hey, we also run into the issue (undefined method `aasm_state') and found the reason why it happened **in our context**. We didn't have a default state_machine but AASM::StateMachineStore generates a default one for you if you call it for example by using ```Car.aasm.human_event_name(some_attribute)```.

In our case this happend while rendering the view. As you can see we forgot to add the state_machine name ```Car.aasm(:our_state_machine).human_event_name(some_attribute)```. The problem is that ```AASM::StateMachineStore``` internally creates a default one that can't work because we didn't have configured a default one. And since it works like a singleton we also had to restart the server afterwards.
